MovingBlocks/Terasology
# test: reducing logspam

I've never been happy with the `setupLogging` method I [added](https://github.com/Terasology/ModuleTestingEnvironment/pull/37) to MTE. Now that it's a year later and we no longer have the engine-tests/MTE split to worry about, I _might_ understand things well enough to do better.

## Contains

* A logback configuration file to be used by default during module tests. Run `groovyw module refresh` to add this `logback-test.xml` to your workspace.

* A new syntax to exclude log messages using a regular expression.

<details>
<summary><b>[ example filter configuration, before and after ]</b></summary>

**Before**, using logback-classic matchers [as documented](https://logback.qos.ch/manual/filters.html#matcher):
```xml
    <filter class="ch.qos.logback.core.filter.EvaluatorFilter">
      <evaluator>
        <matcher>
          <Name>reflectionsEmpty</Name>
          <regex>given scan urls are empty</regex>
        </matcher>
        <expression>reflectionsEmpty.matches(message)</expression>
      </evaluator>
      <OnMismatch>NEUTRAL</OnMismatch>
      <OnMatch>DENY</OnMatch>
    </filter>
```

**After:**
https://github.com/MovingBlocks/Terasology/blob/de1c104b2bcc6705c83177f4eedc2e2347a2413b/engine/src/main/resources/org/terasology/engine/logback/filter-reflections.xml#L8

</details>

* Refactorings to reduce the duplication of the fiddly bits of log configuration between engine tests, module tests, and production (the PC facade).

### Implementation Goals:
- All test tasks should be able to filter logs with logback.
- Test frameworks like JUnit should be `test*` dependencies.
  + except code in engine-tests/src/main that builds on JUnit APIs for use by external module tests, then:
    * `api` if classes/interfaces from it show up on public method signatures.
    * otherwise `implementation` if it's needed to compile but not publicly visible.
- Only the edges (facade, test runners) should define a logging backend (logback).
  + except the code that uses specific logback classes. Which is multiple things in engine, and logback 1.2 isn't structured in a way that facilitates making complete logback extensions from logback-core alone; you often have to pull in logback-classic to do anything useful. This largely breaks the "let the edges pick their own logging backend" architecture I wanted, but the least we can do is not conflict with any `logback.xml` a downstream consumer wants to set up.

## How to test

Run some engine tests, run some module tests, see if you can figure out how to include or exclude the log messages you want during your test output.

## Outstanding before merging

- [x] #5021. I moved dependencies around enough that git will find conflicts.
- [x] log filter for module tests
- [x] document how module maintainers can configure the filter: https://github.com/MovingBlocks/Terasology/wiki/Testing-Modules
- [x] add archive of full test report to the jenkins build, in case you want to get the non-Jenkins-truncated output

## Include with Merge

- [ ] https://github.com/MovingBlocks/ModuleJteConfig/pull/25
